### PR TITLE
Support for automodel hf

### PR DESCRIPTION
**Pull request by ekmb**
- [x] Any pretrained HF model supported via AutoModel Class could be initialized with get_lm_model() in NeMo - no type checking though as every model has different inputs. To enable the type check, a wrapper similar to BertEncoder should be added. 
- [x] Fixed a bug in caching introduced with AutoTokenizer changes
- [x] Jenkins updated:
    - [x] Megatron test changed to multiGPU test
    - [x] added test for t5 model (no wrapper, wasn't supported before)
    - [x] added test for BERT-like model that failed on model_type param due to unconventional naming schema, TurkuNLP/bert-base-finnish-cased-v1 used for the test

### CODE REVIEW

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py [hunk 1]
```diff
@@ -15,7 +15,6 @@
 import re
 from typing import Optional
 
-from transformers import ALL_PRETRAINED_CONFIG_ARCHIVE_MAP
```
Comment (yzhang123) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:18]: Why do we remove ALL_PRETRAINED_CONFIG_ARCHIVE_MAP? If this doesn't return the list, do you know what does?

File: nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py [hunk 2]
```diff
@@ -27,78 +28,94 @@
...
+            config_class = HUGGINGFACE_MODELS[model_type]["config"]
+            return model_class(config_class.from_json_file(config_file))
+        elif config_dict:
+            config_class = HUGGINGFACE_MODELS[model_type]["config"]
+            return model_class(config=config_class(**config_dict))
+        else:
+            return model_class.from_pretrained(pretrained_model_name)
     else:
-        return model_class.from_pretrained(pretrained_model_name)
+        logging.info('Using HuggingFace AutoModel, typing check is not supported')
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py:100]: change to "model type" is not supported or "model class". Typing check is confusing

File: nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py [hunk 3]
```diff
@@ -27,78 +28,94 @@
...
-        return model_class(config=config_class(**config_dict))
+
+    Returns:
+        BertModule
+    """
+
+    # assumes that model configuration is correct, otherwise will pass on Huggingface"s model instantiation error
+    automodel = AutoEncoder(pretrained_model_name)
+
+    model_type = re.search(r'.modeling_[A-z]*[0-5]?\.', str(automodel.type))[0]
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py:87]: how do you know this will work for all cases?

File: nemo/collections/nlp/modules/common/lm_utils.py [hunk 4]
```diff
@@ -59,31 +59,24 @@ def get_lm_model(
...
-            f'pretrained_model_name needs to be from {get_pretrained_lm_models_list()}, however got {pretrained_model_name}'
-        )
-
     # warning when user passes both configuration dict and file
     if config_dict and config_file:
         logging.warning(
             f"Both config_dict and config_file were found, defaulting to use config_file: {config_file} will be used."
         )
 
-    if pretrained_model_name in get_huggingface_pretrained_lm_models_list():
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/lm_utils.py:74]: why did you remove this check?

File: nemo/collections/nlp/modules/common/lm_utils.py [hunk 5]
```diff
@@ -59,31 +59,24 @@ def get_lm_model(
         Pretrained BertModule
     """
 
-    # check valid model type
-    if not pretrained_model_name or pretrained_model_name not in get_pretrained_lm_models_list():
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/lm_utils.py:63]: why are you removing this check?

File: nemo/collections/nlp/modules/common/tokenizer_utils.py [hunk 6]
```diff
@@ -49,34 +49,25 @@ def get_tokenizer(
         special_tokens: dict of special tokens
         vocab_file: path to vocab file
     """
-    full_huggingface_pretrained_model_list = get_huggingface_pretrained_lm_models_list(include_external=True)
```
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/tokenizer_utils.py:52]: same question, why removing all the checks?

File: nemo/collections/nlp/data/glue_benchmark/glue_benchmark_dataset.py [hunk 7]
```diff
@@ -121,7 +121,7 @@ def __init__(
 
         self.examples = processor.get_dev_examples(data_dir) if evaluate else processor.get_train_examples(data_dir)
         processor_name = type(processor).__name__
-        tokenizer_type = type(tokenizer).__name__
+        tokenizer_type = type(tokenizer.tokenizer).__name__
```
Comment (VahidooX) [inline nemo/collections/nlp/data/glue_benchmark/glue_benchmark_dataset.py:124]: Recommend not to use the tokenizer.tokenizer directly from our code other than from the wrapper itself. Our wrapper is supposed to encapsulate the tokenizer. Such encapsulation enables us to have control and later changes and updates easier. Is there a way to set the tokenizer.__name__ to the value of tokenizer.tokenizer.__name__ somehow in the initialization?

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py [hunk 8]
```diff
@@ -106,28 +106,24 @@ def __init__(
         else:
             self.tokenizer = AUTOTOKENIZER.from_pretrained(pretrained_model_name_or_path=pretrained_model_name)
 
-        self.tokenizer.eos_token = self.tokenizer.sep_token
-        self.tokenizer.bos_token = self.tokenizer.cls_token
-
         special_tokens_dict = {}
-        if unk_token:
+        if self.tokenizer.unk_token is None:
```
Comment (VahidooX) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:110]: "is None" or "is not None"?

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py [hunk 9]
```diff
@@ -71,13 +71,13 @@ def __init__(
...
         pretrained_model_name: str,
         vocab_file: Optional[str] = None,
-        mask_token: Optional[str] = None,
-        bos_token: Optional[str] = None,
-        eos_token: Optional[str] = None,
-        pad_token: Optional[str] = None,
-        sep_token: Optional[str] = None,
-        cls_token: Optional[str] = None,
-        unk_token: Optional[str] = None,
+        mask_token: Optional[str] = '[MASK]',
```
Comment (VahidooX) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:74]: Why we set the default values for special tokens? Don't we want user to get error when she tried to use a special token when it is not set or exists for a specific model?

File: nemo/collections/nlp/models/language_modeling/bert_lm_model.py [hunk 10]
```diff
@@ -69,7 +69,7 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
             checkpoint_file=cfg.language_model.lm_checkpoint,
         )
 
-        self.hidden_size = self.bert_model.hidden_size
+        self.hidden_size = self.bert_model.config.hidden_size
```
Comment (VahidooX) [inline nemo/collections/nlp/models/language_modeling/bert_lm_model.py:72]: This hidden_size looks like to just be used in the next line. So how about just use the self.bert_model.config.hidden_size directly or at least not to store it in the self as it is not used?

File: nemo/collections/nlp/modules/common/__init__.py [hunk 11]
```diff
@@ -21,7 +21,11 @@
     DistilBertEncoder,
     RobertaEncoder,
 )
-from nemo.collections.nlp.modules.common.lm_utils import get_lm_model, get_pretrained_lm_models_list
+from nemo.collections.nlp.modules.common.lm_utils import (
+    get_lm_model,
+    get_megatron_lm_models_list,
+    get_pretrained_lm_models_list,
```
Comment (VahidooX) [inline nemo/collections/nlp/modules/common/__init__.py:27]: Is get_pretrained_lm_models_list for HF and get_megatron_lm_models_list for Megatron? If yes, how about naming get_pretrained_lm_models_list to get_hf_lm_models_list?

File: nemo/collections/nlp/modules/common/huggingface/auto.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+from transformers import AutoModel, PreTrainedModel
+
+from nemo.collections.nlp.modules.common.bert_module import BertModule
+from nemo.utils.decorators import experimental
+
+__all__ = ['AutoEncoder']
+
+
+@experimental
+class AutoEncoder(PreTrainedModel, BertModule):
```
Comment (VahidooX) [inline nemo/collections/nlp/modules/common/huggingface/auto.py:26]: AutoEncoder has a special meaning in ML. Lets not to use this name.

File: nemo/collections/nlp/modules/common/megatron/megatron_bert.py [hunk 13]
```diff
@@ -41,19 +42,16 @@ class MegatronBertEncoder(BertModule):
         tokenizer_type (str): tokenizer type, currently only 'BertWordPieceLowerCase' supported.
     """
 
-    def __init__(self, model_name, config, vocab_file):
+    def __init__(self, config, vocab_file):
```
Comment (VahidooX) [inline nemo/collections/nlp/modules/common/megatron/megatron_bert.py:45]: Do we just support BERT version of Megatron or we have have plans for GPT versions too?

File: nemo/collections/nlp/modules/common/megatron/megatron_bert.py [hunk 14]
```diff
@@ -66,13 +64,14 @@ def __init__(self, model_name, config, vocab_file):
         )
 
         # read Megatron arguments back
-        args = get_args()
+        get_args()
```
Comment (VahidooX) [inline nemo/collections/nlp/modules/common/megatron/megatron_bert.py:67]: What does it do?

File: nemo/collections/nlp/modules/common/huggingface/auto.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        PreTrainedModel.__init__(self, config=lm_model.config)
+        self.lm_model = lm_model
+        self.type = type(lm_model)
+
+    def forward(self, **kwargs):
+        unexpected_keys = set(kwargs.keys()) - set(self.lm_model.forward.__code__.co_varnames)
+
+        for key in unexpected_keys:
+            del kwargs[key]
+        res = self.lm_model.forward(**kwargs)[0]
```
Comment (VahidooX) [inline nemo/collections/nlp/modules/common/huggingface/auto.py:43]: What happens to other outputs of the forward?

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py (hunk 9 above)
Comment (ekmb) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:74]: no, we're setting defaults for preprocessing scripts to work, otherwise during preprocessing we won't bet an error, but 'None' will be added instead of token_ids. User always ca overwrite the defaults.

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py (hunk 8 above)
Comment (ekmb) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:110]: yes, we're making sure that the default tokenizer's special tokens are preserved

File: nemo/collections/nlp/data/glue_benchmark/glue_benchmark_dataset.py (hunk 7 above)
Comment (ekmb) [reply nemo/collections/nlp/data/glue_benchmark/glue_benchmark_dataset.py:124]: done

File: nemo/collections/nlp/models/language_modeling/bert_lm_model.py (hunk 10 above)
Comment (ekmb) [reply nemo/collections/nlp/models/language_modeling/bert_lm_model.py:72]: done

File: nemo/collections/nlp/modules/common/__init__.py (hunk 11 above)
Comment (ekmb) [reply nemo/collections/nlp/modules/common/__init__.py:27]: get_pretrained_lm_models_list this includes both HF and megatron

File: nemo/collections/nlp/modules/common/huggingface/auto.py (hunk 12 above)
Comment (ekmb) [reply nemo/collections/nlp/modules/common/huggingface/auto.py:26]: renamed to AutoModelEncoder

File: nemo/collections/nlp/modules/common/huggingface/auto.py (hunk 15 above)
Comment (ekmb) [reply nemo/collections/nlp/modules/common/huggingface/auto.py:43]: others will be passed, just some models don't take say token_type_ids and unless we filter them out, it will cause errors. to handle this better we need a wrapper for each specific model

File: nemo/collections/nlp/modules/common/lm_utils.py (hunk 4 above)
Comment (ekmb) [reply nemo/collections/nlp/modules/common/lm_utils.py:74]: reverted

File: nemo/collections/nlp/modules/common/megatron/megatron_bert.py (hunk 13 above)
Comment (ekmb) [reply nemo/collections/nlp/modules/common/megatron/megatron_bert.py:45]: only BERT for now

File: nemo/collections/nlp/modules/common/megatron/megatron_bert.py (hunk 14 above)
Comment (ekmb) [reply nemo/collections/nlp/modules/common/megatron/megatron_bert.py:67]: It initializes Megatron arguments with megatron module

File: nemo/collections/nlp/modules/common/tokenizer_utils.py (hunk 6 above)
Comment (ekmb) [reply nemo/collections/nlp/modules/common/tokenizer_utils.py:52]: updated

File: nemo/collections/nlp/modules/common/huggingface/auto.py (hunk 15 above)
Comment (VahidooX) [reply nemo/collections/nlp/modules/common/huggingface/auto.py:43]: I mean here (**kwargs)[0] which just selects the 0 output. Don't we need other outputs?

**Review by VahidooX — APPROVED** [review]

**Review by yzhang123 — APPROVED** [review]

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py [hunk 16]
```diff
@@ -95,39 +95,48 @@ def __init__(
...
+            else:
+                self.tokenizer = AUTOTOKENIZER.from_pretrained(pretrained_model_name_or_path=pretrained_model_name)
+        except Exception as e:
+            raise ValueError(f'{pretrained_model_name} is not supported by HuggingFace. {e}')
 
         special_tokens_dict = {}
-        if unk_token:
+        if self.tokenizer.unk_token is None:
             special_tokens_dict["unk_token"] = unk_token
-        if sep_token:
```
Comment (yzhang123) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:115]: @ekmb just realized that now you are not able to set your own custom special tokens anymore please fix that, thanks

Comment (ekmb) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:115]: @yzhang123 what exactly do you want me to fix? you wanted to use eos_token for sep_token when available, no?  
Users can add any special tokens by calling add_special_tokens() and including "'additional_special_tokens'" to the dict of the special tokens after the tokenizer is initialized - the same way it's done in HF.

Comment (yzhang123) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:115]: @ekmb if you pass now special tokens to the constructor it wont work, that is not ok. The constructor should always work
e.g. if you pass sep_token="<<<sep>>>" it won't be updated if the tokenizer.tokenizer.sep_token is set
